cmd/virt-chroot/main.go

```diff
@@ -200,10 +248,53 @@ func main() {
 		createTapCmd,
 		createMDEVCmd,
 		removeMDEVCmd,
+		cgroupsV2DeviceCmd,
 	)
 
 	if err := rootCmd.Execute(); err != nil {
 		_, _ = fmt.Fprintln(os.Stderr, err)
 		os.Exit(1)
 	}
 }
+
+func decodeDeviceRules(marshalledRulesHash string) (unmarshalledRules []*devices.Rule, err error) {
+	marshalledRules, err := base64.StdEncoding.DecodeString(marshalledRulesHash)
+	if err != err {
+		return nil, fmt.Errorf("cannot decode marshalled cgroups v2 rules. "+
+			"encoded rules: %s. err: %v", marshalledRulesHash, err)
+	}
+
+	err = json.Unmarshal(marshalledRules, &unmarshalledRules)
+	if err != nil {
+		return nil, fmt.Errorf("cannot unmarshall cgroup v2 rules. "+
+			"marshalled rules: %s. err: %v", marshalledRules, err)
+	}
+
+	return unmarshalledRules, err
+}
+
+func setCgroupDeviceRules(vmiPidFromHostView uint64, cgroupDirPath string, deviceRules []*devices.Rule, isRootless bool) error {
+	if deviceRules == nil || len(deviceRules) == 0 {
+		return nil
+	}
+
+	config := &configs.Cgroup{
+		Path:      cgroup.HostCgroupBasePath,
+		Paths:     map[string]string{"": cgroupDirPath},
+		Resources: &configs.Resources{},
+	}
+
+	cgroupV2Manager, err := runc_fs2.NewManager(config, cgroupDirPath, isRootless)
+	if err != nil {
+		return fmt.Errorf("cannot create cgroups v2 manager from pid %d. err: %v", vmiPidFromHostView, err)
+	}
+
+	err = cgroupV2Manager.Set(&configs.Resources{
```

Comment:
Hey @vasiliy-ul,

Thanks very much for this. You're right, it seems that runc simply overrides all previous device rules.

The interesting thing is that the same behavior exists for v1. Turns out that if you ask for a certain rule they completely disable any previous rules and only then apply the requested one. Since we had so many workarounds and duplicated-reverse-engineered code from runc we weren't affected by their logic (in a price of not maintaining this code + having a logic which is clearly buggy in same corner cases).

To address that with v1 I've added the `addCurrentRules()` function, although as you said it seems more problematic to to that in v2 since you need to somehow re-compile eBPF code into device rules. On the other hand - I find it hard to believe that such behavior is needed only by us - this sounds like a problem that demands a solution broader than Kubevirt's use.

I will try contacting people from runc repo and discuss about this - or even have a meeting to talk about it. There is always a possibility to solve it internally by all sorts of workarounds. For example, we can somehow save mappings from VMI to its previous requested rules. Then when asking for a new rule we can append it to the previous request and override only the previous version of the requested rule. But again, this sounds like a behaviour that many will require, so I think that the right way of doing things is through runc at this point.

Reply:
Let me update that I've created this conversation in runc's dev channel:
https://groups.google.com/a/opencontainers.org/g/dev/c/tCa1feBywPY

Reply:
Seems that their dev channel is not very alive, so I've opened an issue: https://github.com/opencontainers/runc/issues/3196